extensions/amp-3q-player/0.1/amp-3q-player.js

```diff
@@ -86,17 +83,42 @@ class Amp3QPlayer extends AMP.BaseElement {
     Services.videoManagerForDoc(el).register(this);
   }
 
-  /** @override */
-  layoutCallback() {
-    const iframe = createFrameFor(
-      this,
-      'https://playout.3qsdn.com/' +
+  /** private */
+  generateIframeSrc() {
+    const explicitParamsAttributes = ['key', 'timestamp', 'controls'];
+
+    let iframeSrc = 'https://playout.3qsdn.com/';
+
+    if (this.element.getAttribute(`data-datasource`)) {
+      iframeSrc +=
+        'config_by_metadata/' +
+        this.element.getAttribute(`data-project`) +
+        '/' +
+        this.element.getAttribute(`data-datafield`) +
+        '/' +
         encodeURIComponent(dev().assertString(this.dataId)) +
         // Autoplay is handled by VideoManager
-        '?autoplay=false&amp=true'
-    );
+        '?autoplay=false&amp=true';
+    } else {
+      iframeSrc +=
+        this.dataId +
+        // Autoplay is handled by VideoManager
+        '?autoplay=false&amp=true';
+    }
 
-    this.iframe_ = iframe;
+    explicitParamsAttributes.forEach(explicitParam => {
+      const val = this.element.getAttribute(`data-${explicitParam}`);
```

Comment:
You can just use `data-param-*` here like [`amp-video-iframe`](https://amp.dev/documentation/components/amp-video-iframe/). Use `getDataParamsFromAttributes` and you get that out of the box.

I don't think you need an allowlist for attributes if you explicitly set the `-param-` prefix.

Reply:
The url structure changes with some params, i'll think about it.